dynamic-ssl/src/main/java/org/wildfly/security/dynamic/ssl/DynamicSSLContextSpiImpl.java

```diff
@@ -0,0 +1,146 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wildfly.security.dynamic.ssl;
+
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLContextSpi;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLParameters;
+import javax.net.ssl.SSLServerSocketFactory;
+import javax.net.ssl.SSLSessionContext;
+import javax.net.ssl.SSLSocketFactory;
+import javax.net.ssl.TrustManager;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.util.Iterator;
+import java.util.ServiceLoader;
+
+/**
+ * SSLContextSpi that uses ServiceLoader to find implementations of DynamicSSLContextSPI.
+ * DynamicSSLContextSPI implementation is being used to obtain authentication configuration for DynamicSSLContext.
+ * if no provider is found then SSLContext.getDefault() is used.
+ *
+ * @author <a href="mailto:dvilkola@redhat.com">Diana Krepinska (Vilkolakova)</a>
+ */
+final class DynamicSSLContextSpiImpl extends SSLContextSpi {
+
+    private final DynamicSSLContextSPI dynamicSSLContextImpl;
+    private volatile SSLSocketFactory sslSocketFactory;
+
+    DynamicSSLContextSpiImpl() {
+        this(null);
+    }
+
+    DynamicSSLContextSpiImpl(DynamicSSLContextSPI dynamicSSLContextSPIImpl){
+        if (dynamicSSLContextSPIImpl != null) {
+            dynamicSSLContextImpl = dynamicSSLContextSPIImpl;
+        } else {
+            Iterator<DynamicSSLContextSPI> dynamicSSLContextSPIIterator = ServiceLoader.load(DynamicSSLContextSPI.class).iterator();
+            if (dynamicSSLContextSPIIterator.hasNext()) {
+                dynamicSSLContextImpl = dynamicSSLContextSPIIterator.next();
+            } else {
+                dynamicSSLContextImpl = null;
+            }
+        }
+    }
+
+    private SSLContext getConfiguredDefaultSSLContext() {
+        try {
+            if (dynamicSSLContextImpl != null) {
+                SSLContext configuredDefault = dynamicSSLContextImpl.getConfiguredDefault();
+                if (configuredDefault != null) {
+                    return configuredDefault;
+                }
+            }
+            return SSLContext.getDefault();
+        } catch (NoSuchAlgorithmException | DynamicSSLContextException e) {
+            throw new IllegalStateException(ElytronMessages.log.cannotObtainConfiguredDefaultSSLContext());
+        }
+    }
+
+    @Override
+    protected void engineInit(KeyManager[] keyManagers, TrustManager[] trustManagers, SecureRandom secureRandom) {
+        // ignore
```

Comment:
May be good to add a comment here to explain why this is empty.

Reply:
Fixed